hydro_deploy/core/src/ssh.rs

```diff
@@ -75,81 +80,85 @@ impl LaunchedBinary for LaunchedSshBinary {
         }
     }
 
-    async fn wait(&mut self) -> Option<i32> {
+    async fn wait(&mut self) -> Result<i32> {
         self.channel.wait_eof().await.unwrap();
-        let ret = self.exit_code();
+        let exit_code = self.channel.exit_status()?;
         self.channel.wait_close().await.unwrap();
-        ret
+
+        Ok(exit_code)
+    }
+
+    async fn stop(&mut self) -> Result<()> {
+        if !self.channel.eof() {
+            self.channel.write_all(b"\x03").await?; // `^C`
+            self.channel.send_eof().await?;
+            self.channel.wait_eof().await?;
+            // `exit_status()`
+            self.channel.wait_close().await?;
+        }
+
+        // Run perf post-processing and download perf output.
+        if let Some(perf) = &self.perf {
+            let local_file = perf.output_file.to_str().unwrap();
+            let mut local_file = tokio::fs::File::create(local_file).await?;
+            {
+                let mut script_channel = async_retry(
```

Comment:
Hmm I wonder if this retry is necessary; because if the session is already created opening a channel should always succeed and if it fails will not succeed later? Although I may be forgetting where networking happens in this library.

Reply:
I just left it as a matched pattern, Ill do a test run with it taken out